pkg/analyzer/analyzer.go

```diff
@@ -1,25 +1,190 @@
+// stack must be nonempty for any of its methods aside from `PushCurrent` to be called
+type improperDurationContextStack struct {
+	pass     *analysis.Pass
+	sentinel improperDurationContext
+	slice    []improperDurationContext
+}
+
+func (s *improperDurationContextStack) PushCurrent(node ast.Node) {
+	isMultiplicationExpr := false
+	expr, isExpr := node.(ast.Expr)
+	if isExpr {
+		if binaryExpr, isBinaryOp := expr.(*ast.BinaryExpr); isBinaryOp {
+			isMultiplicationExpr = binaryExpr.Op == token.MUL
+		}
+	}
+	isDurationExpr := false
+	if isExpr {
+		isDurationExpr = isDurationType(s.pass.TypesInfo.TypeOf(expr))
+	}
+	s.slice = append(s.slice, improperDurationContext{
+		node:                 node,
+		isMultiplicationExpr: isMultiplicationExpr,
+		isDurationType:       isDurationExpr,
+	})
+}
+
+func (s *improperDurationContextStack) PopCurrent() {
+	current := s.current()
+	parent := s.parent()
+	isImproperDuration := current.isImproperDuration()
+	isProperDuration := current.isProperDuration()
+	if isImproperDuration {
+		parent.hasImproperDurationChild = true
+	}
+	if isProperDuration {
+		parent.hasProperDurationChild = true
+	}
+	switch {
+	case isProperDuration:
+		// nothing; it's been resolved
+	case !current.isMultiplicationExpr:
+		// immediately report all if any errors; this is not multiplication so it's too late to
+		// fix conversions without a time unit, e.g. `(time.Duration(5) + time.Second)`
+		for _, diagnostic := range current.deferredImproperDurationDiagnostics {
+			s.pass.Report(*diagnostic)
+		}
+	default:
+		// parent might still be able to resolve these improper durations via multiplication with
+		// a proper duration, e.g. `(5 * time.Duration(someInt)) * time.Second`
+		parent.deferredImproperDurationDiagnostics = append(
+			parent.deferredImproperDurationDiagnostics,
+			current.deferredImproperDurationDiagnostics...,
+		)
+	}
+	// free up memory now that diagnostics are deferred or reported
+	current.deferredImproperDurationDiagnostics = nil
+	s.slice = s.slice[:len(s.slice)-1]
+}
+
+func (s *improperDurationContextStack) current() *improperDurationContext {
+	return &s.slice[len(s.slice)-1]
+}
+
+func (s *improperDurationContextStack) parent() *improperDurationContext {
+	if len(s.slice) == 1 {
+		return &s.sentinel
+	}
+	return &s.slice[len(s.slice)-2]
+}
+
+func (s *improperDurationContextStack) ReportImproperDurationNode(diagnostic *analysis.Diagnostic) {
+	s.current().isImproperDurationNode = true
+	// parent might be able to resolve this improper duration via multiplication with a
+	// proper duration, e.g. parent could be `time.Duration(5) * time.Second` when child is
+	// `time.Duration(5)`
+	parent := s.parent()
+	parent.deferredImproperDurationDiagnostics = append(
+		parent.deferredImproperDurationDiagnostics,
+		diagnostic,
+	)
+}
+
+func (s *improperDurationContextStack) Finish() {
+	for _, diagnostic := range s.sentinel.deferredImproperDurationDiagnostics {
+		s.ReportImproperDurationNode(diagnostic)
```

Comment:
This should be `s.pass.Report(diagnostic)`. Although sentinel in practice will never have diagnostics.